python/lib/core/dmod/core/common/collections/catalog.py

```diff
@@ -0,0 +1,651 @@
+    """
+    def to_json(self, *args, **kwargs) -> str:
+        ...
+
+
+T = typing.TypeVar("T")
+"""Any sort of class that might indicate consistency"""
+
+HashableTypeT = typing.TypeVar(
+    "HashableTypeT",
+    bound=typing.Union[
+        typing.Mapping[
+            str,
+            typing.Union[
+                typing.Hashable,
+                typing.Mapping[str, typing.ForwardRef("HashableType")],
+                typing.Iterable[typing.ForwardRef("HashableType")]
+            ]
+        ],
+        typing.Iterable[
+            typing.Union[
+                typing.Hashable,
+                typing.Mapping[
+                    str,
+                    typing.Union[
+                        typing.Hashable,
+                        typing.Mapping[str, typing.ForwardRef("HashableType")],
+                        typing.Iterable[typing.ForwardRef("HashableType")]
+                    ]
+                ],
+                typing.Iterable[typing.ForwardRef("HashableType")]
+            ]
+        ],
+        ToJsonProtocol,
+        ToDictProtocol,
+        typing.Hashable,
+        None
+    ]
+)
+"""A type of item that may either be hashed or we have a method of hashing (such as `hash_hashable_map_sequence`)"""
+
+
+def hash_hashable_map_sequence(value: HashableTypeT) -> int:
+    """
+    Generate the hash or hash of a collection of objects
+
+    Args:
+        value: The value to be hashed
+
+    Returns:
+        The result of the hashing operation
+    """
+    def key_function(element: T) -> int:
+        """
+        Function used to provided a value used to act as the key value for sorting
+
+        Args:
+            element: The value serving as the initial key for sorting
+
+        Returns:
+            A representation of that key value that may be used for comparisons
+        """
+        if isinstance(element, typing.Hashable):
+            return hash(element)
+
+        return hash(str(element))
+
+    if isinstance(value, typing.Mapping):
+        return hash(
+            tuple(
+                (
+                    key if isinstance(key, typing.Hashable) else hash_hashable_map_sequence(key),
+                    value if isinstance(value, typing.Hashable) else hash_hashable_map_sequence(value)
+                )
+                for key, value in sorted(value.items(), key=key_function)
+            )
+        )
+
+    if not isinstance(value, (str, bytes)) and isinstance(value, typing.Iterable):
+        return hash(
+            tuple(
+                hash_hashable_map_sequence(item) for item in value
+            )
+        )
+
+    if isinstance(value, ToDictProtocol):
+        return hash_hashable_map_sequence(value.to_dict())
+
+    if isinstance(value, ToJsonProtocol):
+        return hash_hashable_map_sequence(value.to_json())
+
+    return hash(value)
+
+
+class CatalogEntry(typing.Generic[HashableTypeT]):
```

Comment:
The more I think about it, the more I am convinced that we could use a `multiprocessing.shared_memory.SharedMemory` object in place of this class. It would generate the identifier and it has data. It is going to be tricky to handle `last_accessed` in anycase. I have a feeling if we _need_ to support `last_accessed` we will have to get creative.

Reply:
If the _data_ behind an entry _needs_ to be mutable, this will need some kind of locking mechanism. Off the top of my head an `RWLock` would make the most sense. However, I would strongly advise that we avoid allowing mutation and that items are strictly immutable once they are in the cache. This might not be feasible, but we should push that thought until we are certain we _need_ cache item mutability. The problem is far simpler if mutations are disallowed.